main/src/ca/uwaterloo/flix/api/lsp/provider/GotoProvider.scala

```diff
@@ -27,74 +29,129 @@ object GotoProvider {
-        case Entity.Op(_) => mkNotFound(uri, pos)
-        case Entity.Pred(_, _) => mkNotFound(uri, pos)
-        case Entity.Sig(_) => mkNotFound(uri, pos)
-        case Entity.TypeVar(_) => mkNotFound(uri, pos)
-      }
+  def processGoto(uri: String, pos: Position)(implicit root: Root): JObject = {
+    val gotoRight = searchRight(uri, pos).flatMap(goto)
+    val gotoLeft = searchLeft(uri, pos).flatMap(goto)
+
+    gotoRight
+      .orElse(gotoLeft)
+      .getOrElse(mkNotFound(uri, pos))
+  }
+
+  private def searchRight(uri: String, pos: Position)(implicit root: Root): Option[AnyRef] = search(uri, pos)
+
+  private def searchLeft(uri: String, pos: Position)(implicit root: Root): Option[AnyRef] = {
+    if (pos.character >= 2) {
+      val left = Position(pos.line, pos.character - 1)
+      search(uri, left)
+    } else {
+      None
     }
   }
 
+  private def search(uri: String, pos: Position)(implicit root: Root): Option[AnyRef] = {
+    val consumer = StackConsumer();
+    Visitor.visitRoot(root, consumer, Visitor.InsideAcceptor(uri, pos))
+    consumer.getStack.filter(isReal).headOption
+  }
+
+  private def goto(x: AnyRef)(implicit root: Root): Option[JObject] = x match {
+    // Assoc Types
+    case SymUse.AssocTypeSymUse(sym, loc) => Some(mkGoto(LocationLink.fromAssocTypeSym(sym, loc)))
+    case AssocTypeConstructor(sym, loc) => Some(mkGoto(LocationLink.fromAssocTypeSym(sym, loc)))
+    case Type.AssocType(AssocTypeConstructor(sym, _), _, _, loc) => Some(mkGoto(LocationLink.fromAssocTypeSym(sym, loc)))
+    // Defs
+    case SymUse.DefSymUse(sym, loc) => Some(mkGoto(LocationLink.fromDefSym(sym, loc)))
+    // Effects
+    case SymUse.EffectSymUse(sym, loc) => Some(mkGoto(LocationLink.fromEffectSym(sym, loc)))
+    case Type.Cst(TypeConstructor.Effect(sym), loc) => Some(mkGoto(LocationLink.fromEffectSym(sym, loc)))
+    case SymUse.OpSymUse(sym, loc) => Some(mkGoto(LocationLink.fromOpSym(sym, loc)))
+    // Enums
+    case Type.Cst(TypeConstructor.Enum(sym, _), loc) => Some(mkGoto(LocationLink.fromEnumSym(sym, loc)))
+    case SymUse.CaseSymUse(sym, loc) => Some(mkGoto(LocationLink.fromCaseSym(sym, loc)))
+    // Struct
+    case Type.Cst(TypeConstructor.Struct(sym, _), loc) => Some(mkGoto(LocationLink.fromStructSym(sym, loc)))
+    case SymUse.StructFieldSymUse(sym, loc) => Some(mkGoto(LocationLink.fromStructFieldSym(sym, loc)))
+    // Traits
+    case SymUse.TraitSymUse(sym, loc) => Some(mkGoto(LocationLink.fromTraitSym(sym, loc)))
+    case TraitConstraint.Head(sym, loc) => Some(mkGoto(LocationLink.fromTraitSym(sym, loc)))
+    case SymUse.SigSymUse(sym, loc) => Some(mkGoto(LocationLink.fromSigSym(sym, loc)))
+    // Type Vars
+    case Type.Var(sym, loc) => Some(mkGoto(LocationLink.fromTypeVarSym(sym, loc)))
+    // Vars
+    case TypedAst.Expr.Var(sym, _, loc) => Some(mkGoto(LocationLink.fromVarSym(sym, loc)))
+    case _ => None
+  }
+
+  private def isReal(x: AnyRef): Boolean = x match {
```

Comment:
Is this function similar to the one from the other provider? Or does it have some subtle differences?

Reply:
I might not literally be the same, but it should be. It's really just answering the question: does this "thing" `x` have a real sourcelocation.

Reply:
Yes, we could extract it from both and have a single definition somewhere. Not sure where it should be though. Maybe it's own file somewhere in LSP directory?

Reply:
If `x` was just AST nodes,  I would say put it in TypedAst.scala, but it's not. It's also SymUse and such. Really, it's anything that the Consumer might consume.

Reply:
Since that change will refactor both providers, I think I'll do that in another PR :)